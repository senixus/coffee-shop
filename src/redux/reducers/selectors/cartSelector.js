import { createSelector } from "reselect";

export const selectCart = (state) => state.cart;

export const selectCartTotalPrice = createSelector([selectCart], (cart) => {
  return cart.reduce(
    (acc, cartItem) => acc + cartItem.quantity * cartItem.price,
    0
  );
});
