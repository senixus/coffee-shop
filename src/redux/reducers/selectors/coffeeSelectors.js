import { createSelector } from "reselect";

const selectAllCoffee = (state) => state.coffee.coffee;

export const selectCoffee = createSelector([selectAllCoffee], (coffee) => {
  return coffee;
});

export const selectLowestPrice = (action) =>
  createSelector([selectCoffee], (coffee) => {
    if (action === "lowest price") {
      return coffee.sort((a, b) => a.price - b.price);
    }
  });

export const selectHighestPrice = (action) =>
  createSelector([selectCoffee], (coffee) => {
    if (action === "highest price") {
      return coffee.sort((a, b) => a.price - b.price).reverse();
    }
  });

export const selectAlphabetic = (action) =>
  createSelector([selectCoffee], (coffee) => {
    if (action === "name") {
      return coffee.sort((a, b) => a.coffeeName.localeCompare(b.coffeeName));
    }
  });

export const selectReverseAlphabetic = (action) =>
  createSelector([selectCoffee], (coffee) => {
    if (action === "name reverse") {
      return coffee
        .sort((a, b) => a.coffeeName.localeCompare(b.coffeeName))
        .reverse();
    }
  });

export const selectSearchCoffee = (action) =>
  createSelector([selectCoffee], (coffee) => {
    return coffee.filter((c) =>
      c.coffeeName.toLowerCase().includes(action.toLowerCase())
    );
  });
