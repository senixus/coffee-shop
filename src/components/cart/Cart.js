import React from "react";
import { useDispatch, useSelector } from "react-redux";
import ProductHeader from "../productHeader/ProductHeader";
import "./cart.scss";
import { IoMdClose } from "react-icons/io";
import {
  removeItemFromCart,
  clearCartItems,
} from "../../redux/actions/cart/removeFromCart";

import { addToCart } from "../../redux/actions/cart/addToCart";
import { selectCartTotalPrice } from "../../redux/reducers/selectors/cartSelector";
import CheckoutButton from "../checkoutButton/CheckoutButton";

const Cart = () => {
  const cart = useSelector((state) => state.cart.cart);
  const dispatch = useDispatch();
  const removeCart = (coffee) => dispatch(removeItemFromCart(coffee));
  const total = useSelector((state) => selectCartTotalPrice(state.cart));
  const addCart = (coffee) => dispatch(addToCart(coffee));
  const clearItemFromCart = (coffee) => dispatch(clearCartItems(coffee));

  return (
    <main className="cart">
      <ProductHeader title="Cart Detail" />

      <div className="cart-child container">
        {cart &&
          cart.map((coffee) => (
            <table className="cart-info" key={coffee.id}>
              <thead>
                <tr>
                  <th></th>
                  <th>Product</th>
                  <th>Quantity</th>
                  <th>Total</th>
                </tr>
              </thead>
              <tbody>
                <tr>
                  <td>
                    <button
                      className="close-btn"
                      onClick={() => clearItemFromCart(coffee)}
                    >
                      <IoMdClose />
                    </button>
                  </td>
                  <td>
                    <span>
                      <img
                        src={coffee.coffeePic.name}
                        alt={coffee.coffeeName}
                      />
                    </span>
                    <span style={{ marginLeft: ".4rem" }}>
                      {coffee.coffeeName}
                    </span>
                  </td>
                  <td>
                    <span className="arrows" onClick={() => removeCart(coffee)}>
                      &#10094;
                    </span>
                    {coffee.quantity}
                    <span className="arrows" onClick={() => addCart(coffee)}>
                      &#10095;
                    </span>
                  </td>
                  <td>$ {coffee.price.toFixed(2)}</td>
                </tr>
              </tbody>
            </table>
          ))}

        <div className="cart-totals">
          <table>
            <tbody>
              <tr className="cart-subtotal">
                <th>Total : </th>
                <td>$ {total.toFixed(2)}</td>
              </tr>
            </tbody>
          </table>
          <div style={{ color: "red", fontSize: "1.3rem" }}>
            <h4>Test Card Number : </h4>
            <p>4242 4242 4242 4242</p>
            <p>10-22 | 123</p>
          </div>
          <div className="cart-totals__checkout">
            <CheckoutButton />
          </div>
        </div>
      </div>
    </main>
  );
};

export default Cart;
