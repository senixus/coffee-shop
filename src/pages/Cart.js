import Cart from "../components/cart/Cart";
import ReactHelmet from "../components/helmet/ReactHelmet";
const CartPage = () => {
  return (
    <>
      <ReactHelmet title="Cart" />
      <Cart />
    </>
  );
};

export default CartPage;
