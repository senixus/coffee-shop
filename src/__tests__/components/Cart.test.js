import React from "react";
//import { shallow } from "enzyme";
import Cart from "../../components/cart/Cart";
//import * as redux from "react-redux";
import { selectCartTotalPrice } from "../../redux/reducers/selectors/cartSelector";
//import * as actionTypes from "../../redux/actions/actionTypes";
import { mockStore } from "../../setupTests";
//import { removeItemFromCart } from "../../redux/actions/cart/removeFromCart";
import renderer from "react-test-renderer";
import { Provider } from "react-redux";
import { BrowserRouter } from "react-router-dom";

jest.mock("../../redux/reducers/selectors/cartSelector");

// jest.mock("../../redux/actions/cart/removeFromCart.js", () => ({
//   removeItemFromCart: jest.fn(),
// }));
console.log("Something");

const cart = [
  {
    id: 1,
    coffeeName: "Test product",
    quantity: 1,
    price: 16,
    coffeePic: { name: "image" },
  },
];

// const mockState = {
//   cart: cart,
//   total: selectCartTotalPrice(cart),
// };

describe("Cart component", () => {
  // let spyOnUseSelector;
  // let spyOnUseDispatch;
  // let mockDispatch;
  let store;
  let component;
  beforeEach(() => {
    store = mockStore({
      cart: cart,
      total: selectCartTotalPrice(cart),
      auth: {
        user: { uid: 1, name: "Test user" },
      },
    });

    component = renderer.create(
      <BrowserRouter>
        {" "}
        <Provider store={store}>
          <Cart />
        </Provider>
      </BrowserRouter>
    );

    // // Mock useSelector hook
    // spyOnUseSelector = jest
    //   .spyOn(redux, "useSelector")
    //   .mockImplementation((cb) => cb(mockState));
    // // spyOnUseSelector.mockReturnValue();

    // // Mock useDispatch hook
    // spyOnUseDispatch = jest.spyOn(redux, "useDispatch");
    // // Mock dispatch function returned from useDispatch
    // mockDispatch = jest.fn();

    // spyOnUseDispatch.mockReturnValue(mockDispatch);
  });

  afterEach(() => {
    jest.restoreAllMocks();
  });

  it("expect to render Cart component", () => {
    // const wrapper = shallow(<Cart />);
    // expect(wrapper.exists()).toBe(true);
    // expect(wrapper.length).toBe(1);
    // expect(wrapper).toMatchSnapshot();
    expect(component.toJSON()).toMatchSnapshot();
  });
  //   it("expect to render something", () => {
  //     //console.log(component.root.findAllByType("span")[4].props);

  //     renderer.act(() => {
  //       component.root.findAllByType("span")[2].props.onClick();
  //     });

  //     expect(store.dispatch).toHaveBeenCalledTimes(1);
  //     expect(store.dispatch).toHaveBeenCalledWith(
  //       removeItemFromCart({
  //         id: 1,
  //         coffeeName: "Test product",
  //         quantity: 1,
  //         price: 16,
  //         coffeePic: { name: "image" },
  //       })
  //     );
  //   });
});
