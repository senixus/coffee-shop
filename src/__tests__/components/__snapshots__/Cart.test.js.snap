// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cart component expect to render Cart component 1`] = `
Array [
  " ",
  <main
    className="cart"
  >
    <header
      className="product-header"
    >
      <nav
        className=" navbar "
      >
        <div
          className="navbar-logo"
        >
          <h1>
            <a
              className="navbar-logo__item"
              href="/"
              onClick={[Function]}
            >
              <img
                alt="Logo"
                src="coffee.png"
                width="70px"
              />
            </a>
          </h1>
        </div>
        <ul
          className="navbar-nav"
        >
          <li
            className="navbar-nav__item"
          >
            <a
              className="navbar-nav__link"
              href="/"
              onClick={[Function]}
            >
              Home
            </a>
          </li>
          <li
            className="navbar-nav__item"
          >
            <a
              className="navbar-nav__link"
              href="/shop"
              onClick={[Function]}
            >
              Shop
            </a>
          </li>
          <li
            className="navbar-nav__item"
          >
            <a
              className="navbar-nav__link"
              href="/contact"
              onClick={[Function]}
            >
              Contact
            </a>
          </li>
          <li
            className="navbar-nav__item cart"
          >
            <a
              className="navbar-nav__link "
              href="/cart"
              onClick={[Function]}
            >
              <svg
                fill="currentColor"
                height="1em"
                stroke="currentColor"
                strokeWidth="0"
                style={
                  Object {
                    "color": undefined,
                  }
                }
                viewBox="0 0 576 512"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M528.12 301.319l47.273-208C578.806 78.301 567.391 64 551.99 64H159.208l-9.166-44.81C147.758 8.021 137.93 0 126.529 0H24C10.745 0 0 10.745 0 24v16c0 13.255 10.745 24 24 24h69.883l70.248 343.435C147.325 417.1 136 435.222 136 456c0 30.928 25.072 56 56 56s56-25.072 56-56c0-15.674-6.447-29.835-16.824-40h209.647C430.447 426.165 424 440.326 424 456c0 30.928 25.072 56 56 56s56-25.072 56-56c0-22.172-12.888-41.332-31.579-50.405l5.517-24.276c3.413-15.018-8.002-29.319-23.403-29.319H218.117l-6.545-32h293.145c11.206 0 20.92-7.754 23.403-18.681z"
                />
              </svg>
              <span
                className="basket-count"
              />
            </a>
            <div
              className="dropdown"
            >
              <div
                className="dropdown-info"
              >
                <div
                  className="total-price"
                >
                  <p>
                    TOTAL
                  </p>
                  <p>
                    $
                  </p>
                </div>
                <div
                  className="dropdown-info__btn"
                >
                  <a
                    className="view-btn"
                    href="/cart"
                    onClick={[Function]}
                  >
                    view cart
                  </a>
                </div>
              </div>
            </div>
          </li>
          <li
            className="navbar-nav__item profile"
          >
            <a
              className="navbar-nav__link"
              href="/"
              onClick={[Function]}
            />
            <ul
              className="profile-dropdown"
            >
              <li>
                <a
                  href="/wishlist"
                  onClick={[Function]}
                >
                  <span
                    className="icon"
                  >
                    <svg
                      fill="none"
                      height="1em"
                      stroke="currentColor"
                      strokeLinecap="round"
                      strokeLinejoin="round"
                      strokeWidth="2"
                      style={
                        Object {
                          "color": undefined,
                        }
                      }
                      viewBox="0 0 24 24"
                      width="1em"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <line
                        x1="8"
                        x2="21"
                        y1="6"
                        y2="6"
                      />
                      <line
                        x1="8"
                        x2="21"
                        y1="12"
                        y2="12"
                      />
                      <line
                        x1="8"
                        x2="21"
                        y1="18"
                        y2="18"
                      />
                      <line
                        x1="3"
                        x2="3.01"
                        y1="6"
                        y2="6"
                      />
                      <line
                        x1="3"
                        x2="3.01"
                        y1="12"
                        y2="12"
                      />
                      <line
                        x1="3"
                        x2="3.01"
                        y1="18"
                        y2="18"
                      />
                    </svg>
                  </span>
                  Wish List
                </a>
              </li>
              <li>
                <a
                  href="/"
                  onClick={[Function]}
                >
                  <span
                    className="icon"
                  >
                    <svg
                      fill="none"
                      height="1em"
                      stroke="currentColor"
                      strokeLinecap="round"
                      strokeLinejoin="round"
                      strokeWidth="2"
                      style={
                        Object {
                          "color": undefined,
                        }
                      }
                      viewBox="0 0 24 24"
                      width="1em"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        d="M9 21H5a2 2 0 0 1-2-2V5a2 2 0 0 1 2-2h4"
                      />
                      <polyline
                        points="16 17 21 12 16 7"
                      />
                      <line
                        x1="21"
                        x2="9"
                        y1="12"
                        y2="12"
                      />
                    </svg>
                  </span>
                  Logout
                </a>
              </li>
            </ul>
          </li>
          <li
            className="bar"
            onClick={[Function]}
          >
            <svg
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              strokeWidth="0"
              style={
                Object {
                  "color": undefined,
                }
              }
              viewBox="0 0 448 512"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M16 132h416c8.837 0 16-7.163 16-16V76c0-8.837-7.163-16-16-16H16C7.163 60 0 67.163 0 76v40c0 8.837 7.163 16 16 16zm0 160h416c8.837 0 16-7.163 16-16v-40c0-8.837-7.163-16-16-16H16c-8.837 0-16 7.163-16 16v40c0 8.837 7.163 16 16 16zm0 160h416c8.837 0 16-7.163 16-16v-40c0-8.837-7.163-16-16-16H16c-8.837 0-16 7.163-16 16v40c0 8.837 7.163 16 16 16z"
              />
            </svg>
          </li>
        </ul>
      </nav>
      <div
        className="product-header__child"
      >
        <h1>
          Cart Detail
        </h1>
      </div>
    </header>
    <div
      className="cart-child container"
    >
      <div
        className="cart-totals"
      >
        <table>
          <tbody>
            <tr
              className="cart-subtotal"
            >
              <th>
                Total : 
              </th>
              <td>
                $ 
              </td>
            </tr>
          </tbody>
        </table>
        <div
          style={
            Object {
              "color": "red",
              "fontSize": "1.3rem",
            }
          }
        >
          <h4>
            Test Card Number : 
          </h4>
          <p>
            4242 4242 4242 4242
          </p>
          <p>
            10-22 | 123
          </p>
        </div>
        <div
          className="cart-totals__checkout"
        >
          <div
            onMouseEnter={[Function]}
            onMouseLeave={[Function]}
            style={
              Object {
                "position": "fixed",
                "zIndex": 9999,
              }
            }
          />
          <button
            className="StripeCheckout"
            onBlur={[Function]}
            onClick={[Function]}
            onFocus={[Function]}
            onMouseDown={[Function]}
            onMouseOut={[Function]}
            onMouseUp={[Function]}
            style={
              Object {
                "background": "linear-gradient(#28a0e5,#015e94)",
                "border": 0,
                "borderRadius": 5,
                "boxShadow": "0 1px 0 rgba(0,0,0,0.2)",
                "cursor": "pointer",
                "display": "inline-block",
                "overflow": "hidden",
                "padding": 1,
                "textDecoration": "none",
                "userSelect": "none",
                "visibility": "visible",
              }
            }
          >
            <span
              style={
                Object {
                  "backgroundImage": "linear-gradient(#7dc5ee,#008cdd 85%,#30a2e4)",
                  "borderRadius": 4,
                  "boxShadow": "inset 0 1px 0 rgba(255,255,255,0.25)",
                  "color": "#fff",
                  "display": "block",
                  "fontFamily": "\\"Helvetica Neue\\",Helvetica,Arial,sans-serif",
                  "fontSize": 14,
                  "fontWeight": "bold",
                  "height": 30,
                  "lineHeight": "30px",
                  "padding": "0 12px",
                  "position": "relative",
                  "textShadow": "0 -1px 0 rgba(0,0,0,0.25)",
                }
              }
            >
              Pay Now
            </span>
          </button>
        </div>
      </div>
    </div>
  </main>,
]
`;
